# Market Analysis: Problem Set 3

**Submission guidelines:**  

1. Submission requires downloading and saving your ipynb file as .html file with the filename, studentID-studentName. For example, 407510073-束良鴻.
2. Removing all irrelevant code blocks.
3. Email the html file to your TA. (E-mail address: p4896p4896@gmail.com)
4. Failing to submit in a correct format will result in a reduction of **20 points**. 
5. Deadline: 4/28 Thur. 1 p.m. 
6. No late submission will be accepted. 

## Preparation

直接執行下面一欄，把所有必備之套件正確引用。 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

請把 **tipsdata.csv** 存成Pandas DataFrame Object，並命名為tips。 

### 本資料為某間餐廳的小費資料，相關欄位有總消費金額(total_bill \\$)，小費金額(tip \\$)，服務員之性別(sex)，有沒有抽菸(smoker)，星期幾(day)，時間(time)，以及顧客人數(size)。

In [3]:
tips

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

## Q1 PANDAS資料分析 (45\%)

(a) 請印出前5行資料。(2 pts)

(b) 請問該資料檔共有幾個觀測值? (2 pts)

(c) 請**隨機**印出11行資料。(你的結果應該會跟上面的不一樣。) (5 pts)

(d) 請問午餐(Lunch)與晚餐(Dinner)的平均消費金額為多少? (8 pts)

(e) 請問顧客人數為4人時，所給的小費平均金額為多少? (8 pts)

(f) 請問有幾個觀測值是有抽菸的? (10 pts)

(g) 請創建一個新欄位，名稱為**percent**，來計算小費佔總消費金額之比例。並找出最高的比例為何? (10 pts)

## Q2 視覺化分析 (25\%)

(a) 請用Seaborn，繪製出消費金額(total_bill \\$)，小費金額(tip \\$)，顧客人數(size)的pairplot。(只有三個數值欄位，所以僅有3$\times$3) (5 pts)

(b) 請試圖畫出一個joint plot，x軸用total_bill，y軸用tip。 (5 pts)

(c) 請畫出total_bill的直方圖。 (5 pts)

(d) 畫出**total_bill**與**tip**的散布圖(scatter plot)，並依照顧客人數決定圓點大小，最後把透明度調整為0.3。 (10 pts)

## Q3 Linear Regression (40\%)

在進行迴歸分析之前，我們要先把一些質性的變數轉換為虛擬變數(dummy variables)，並把它存為新的DataFrame名為tip_lm。請先執行下面程式碼。

In [14]:
tip_lm = pd.concat([tips, pd.get_dummies(tips['sex']), pd.get_dummies(tips['time']), pd.get_dummies(tips['smoker'])], axis=1)
tip_lm

total_bill   tip     sex smoker   day    time  size  Female  Male  \
0         16.99  1.01  Female     No   Sun  Dinner     2       1     0   
1         10.34  1.66    Male     No   Sun  Dinner     3       0     1   
2         21.01  3.50    Male     No   Sun  Dinner     3       0     1   
3         23.68  3.31    Male     No   Sun  Dinner     2       0     1   
4         24.59  3.61  Female     No   Sun  Dinner     4       1     0   
..          ...   ...     ...    ...   ...     ...   ...     ...   ...   
239       29.03  5.92    Male     No   Sat  Dinner     3       0     1   
240       27.18  2.00  Female    Yes   Sat  Dinner     2       1     0   
241       22.67  2.00    Male    Yes   Sat  Dinner     2       0     1   
242       17.82  1.75    Male     No   Sat  Dinner     2       0     1   
243       18.78  3.00  Female     No  Thur  Dinner     2       1     0   

     Dinner  Lunch  No  Yes  
0         1      0   1    0  
1         1      0   1    0  
2         1      0   1    0  
3         1      0   1    0  
4         1      0   1    0  
..      ...    ...  ..  ...  
239       1      0   1    0  
240       1      0   0    1  
241       1      0   0    1  
242       1      0   1    0  
243       1      0   1    0  

[244 rows x 13 columns]

從上面新的DataFrame看來，我們已經把依序把三個質性的變數，轉成兩個虛擬變數sex(Female、Male)，time(Dinner、Lunch)，somoker(Yes、No)。
接著定義你的X變數與Y變數，並且引用線性模型的模組。
(這一題不需使用train_test_split，我們把所有樣本都拿來估計與預測。)

In [18]:
from sklearn.linear_model import LinearRegression
Y = tip_lm['tip']
X = tip_lm[['total_bill', 'Female', 'size', 'Dinner', 'Yes']]

(a) 請用上面之X與Y來訓練或估計一個線性迴歸模型。並該模型之係數為何? (10 pts)

(b) 請用這個模型來預測所有244個樣本的小費金額，並把它存在predictions裡面，再來把predictions給印出來。 (10 pts)

(c) 請將實際小費減去預測小費後，計算出誤差項後，畫出該之直方圖與分布線。 (10 pts)

(d) 請計算出該樣本之MSE(mean squared errors)。 (10 pts)